### PR TITLE
[WIP] use pinned object heap in array pool.

### PR DESCRIPTION
**Pull request by VSadov**

### CODE REVIEW

**Comment by stephentoub** [comment]
What's the rationale for this?

**Comment by VSadov** [comment]
CC: @jkotas @Maoni0

**Comment by VSadov** [comment]
@stephentoub - immediate advantage is that buffers do not need to go Gen0->Gen1->Gen2

**Comment by VSadov** [comment]
@stephentoub - It is also a good assumption that these buffers get pinned and less pinning in the SOH is a good thing. 

The next step, could be avoid/simplify pinning if we know that the array is from the pool.

**Comment by stephentoub** [comment]
Is it as cheap to allocate a pinned array as it is an unpinned one?  There are plenty of situations where we would expect the number of arrays the pool is willing to hold to be less than the number needed at any one time, which means some percentage of pool accesses are likely going to be resulting in allocating rather than returning a cached array.

**Comment by jkotas** [comment]
Agree with Stephen. The pool implementation would need to be changed to be more like malloc/free than just opportunistic pool for this to work well. (IMHO, that is hard to do without making the GC to be aware of the pool existence.)

**Comment by jkotas** [comment]
> Is it as cheap to allocate a pinned array as it is an unpinned one? 

There two ways that allocating pinned array is more expensive:
- The path length of the allocation is longer (it is not a helper written in hand-optimized assembly)
- The pinned heap is treated as Gen2. It means that you need to do the more costly Gen2 GC to collect the garbage in it.

**Comment by VSadov** [comment]
Since @stephentoub asked about perf, and I am also curious - here are some very crude numbers.

Allocating 100000 randomly sized arrays in two extreme scenarios - dropping arrays immediately or retaining then for the duration of the test.

```cs
    private static void AllocateUninitializedPinned()
    {
        System.Console.Write("AllocateUninitializedPinned: ");
        var r = new Random(1234);
        Stopwatch sw = Stopwatch.StartNew();
        for (int i = 0; i < iters; i++)
        {
            int size = r.Next(2, 10000);
            var arr = GC.AllocateUninitializedArray<long>(size, pinned: true);
            Add(arr);   // noop when not retaining, store in a static list when do
        }

        sw.Stop();
        System.Console.WriteLine(sw.ElapsedMilliseconds);
    }
```

=== results.

```
====    not retaining  (pure allocate and drop)

New: 91      < this does not use API, just   `new long[size]` for a baseline.
AllocateNotPinned: 92
AllocateUninitializedNotPinned: 12
AllocatePinned: 231
AllocateUninitializedPinned: 229

====    keeping until after the loop

New: 1895
AllocateNotPinned: 2259
AllocateUninitializedNotPinned: 1962
AllocatePinned: 271
AllocateUninitializedPinned: 263
```

These are very rough numbers on extreme ends. Real numbers are probably somewhere in between and scenario-specific. 
For better measurements, I'd use GC simulator, but that is more involved. This is just what can be measured quickly.

Also note that we did not tune POH at all. It basically uses LOH tunings. That could be suboptimal.

**Comment by VSadov** [comment]
This is for randomly retaining 50% of allocations in a 2047 sized buffer without any clearing. 

As in:
```cs
    private static void Add(long[] arr)
    {
        if (r.Next(100) > 50)
            staticArray[arr.Length & 2047] = arr;
    }
```
That is to make sure there is some fragmentation in the heap. 

=== results:
```
New: 1311
AllocateNotPinned: 1333
AllocateUninitializedNotPinned: 1080
AllocatePinned: 291
AllocateUninitializedPinned: 43
```

I would say pinned allocations look quite competitive unless used for very shortlived objects.

File: src/libraries/System.Private.CoreLib/src/System/Buffers/TlsOverPerCoreLockedStacksArrayPool.cs [hunk 1]
```diff
@@ -135,13 +135,21 @@ public override T[] Rent(int minimumLength)
...
             else
             {
                 // The request was for a size too large for the pool.  Allocate an array of exactly the requested length.
                 // When it's returned to the pool, we'll simply throw it away.
-                buffer = GC.AllocateUninitializedArray<T>(minimumLength);
+#if !MONO
+                buffer = GC.AllocateUninitializedArray<T>(minimumLength, pinned: !RuntimeHelpers.IsReferenceOrContainsReferences<T>());
+#else
+                buffer = GC.AllocateUninitializedArray<T>(minimumLength, pinned: false);
+#endif
```
Comment (MihaZupan) [inline src/libraries/System.Private.CoreLib/src/System/Buffers/TlsOverPerCoreLockedStacksArrayPool.cs]: Isn't it counter-productive to pin the arrays that will be thrown away when returned to the pool?

Comment (VSadov) [reply src/libraries/System.Private.CoreLib/src/System/Buffers/TlsOverPerCoreLockedStacksArrayPool.cs]: Here "pinnig" is just allocating into a separate region of the heap that is never compacted. It does not mean some additional action.
In this particular case the array is very large (2^31 elements ?). If we do not specify "pinned" it will end up in the Large Object Heap, which is very similar to Pinned Object Heap, just permits compaction, which may not be a good idea with such a large object.

The main reason to use "pinned" here is consistency with other cases.

Comment (MihaZupan) [reply src/libraries/System.Private.CoreLib/src/System/Buffers/TlsOverPerCoreLockedStacksArrayPool.cs]: Wouldn't the limit be 2^21?

Comment (VSadov) [reply src/libraries/System.Private.CoreLib/src/System/Buffers/TlsOverPerCoreLockedStacksArrayPool.cs]: Yes 2^21 
(4+17), to be sure. :-)
It is a big array.

**Comment by jkotas** [comment]
> I would say pinned allocations look quite competitive 

What is the peak working set for each case? My guess is that difference is caused by more Gen0 GCs in the unpinned case vs. less Gen2 GCs in the pinned case.

> unless used for very shortlived objects.

Which is the common array pool use.

**Comment by VSadov** [comment]
> What is the peak working set for each case?

I looked at the TaskManager numbers. The sustained working set looks similar.

== AllocateUninitializedNotPinned      (New is very similar )
peak: ~194Mb
active: ~180Mb

== AllocateUninitializedPinned
peak: ~194Mb
active: ~180Mb

It looks like it depends more on how much is retained on average than how it is allocated. 

According to VTune Non-pinned case spends most time in foreground compaction, while pinned case spends noticeable (but not as much) time in background marking.

This is kind of expected - 
SOH allocations are cheap if objects die young. If they survive longer, there is additional costs of compacting/promoting.

POH allocations have longer codepath, but would not have compaction costs. Background mark/sweep seems to be doing good job and we do not run into blocking GC problems. 
That explains why pinned case could be faster end-to-end when objects are mid-sized arrays and there is some survivorship.

**Comment by VSadov** [comment]
I wonder - do we want to discuss/decide this further from the perf angle? 
What would be  the gating scenarios and expectations?

Are we ok with this form the functional point in the first place?

**Comment by saucecontrol** [comment]
> > unless used for very shortlived objects.

> Which is the common array pool use.

They're short-use but long-lived

> Are we ok with this from the functional point in the first place?

FWIW, I'd very much like to see this go in.  I typically rent buffers from the pool for either Interop or HWIntrinsics use, so I almost always end up pinning them.  Would be very nice to avoid managed heap fragmentation as a result.

**Comment by jkotas** [comment]
@saucecontrol Do you have data or examples that demonstrate that you have fragmentation problems with the current ArrayPool ?

**Comment by saucecontrol** [comment]
> Do you have data or examples that demonstrate that you have fragmentation problems with the current ArrayPool?

I don't have anything concrete at hand.

Looking through the BCL code, I can see many instances where a rented array is immediately pinned, but they tend to be for shorter duration.  My own use case tends to involve slightly larger buffers (almost always within the pool's limits, and mostly under the LOH size cutoff) and longer pinning durations.  I use pool arrays as buffers for native image codecs to decode into, and the larger the buffer, generally the longer it takes to fill it -- so the longer it stays pinned.

This leads to general concern over fragmentation when my imaging libraries are used in apps that have frequent GCs.  I can minimize allocations in my libraries to reduce GC frequency, but I can't control what the apps that use them do.

**Comment by jkotas** [comment]
> What would be the gating scenarios and expectations?

We need data on what this makes better and what this makes worse. And then we need to agree that the scenarios that it makes better has to be more important that the scenarios that it makes worse.

I have done some ad-hoc testing. I was not able to come up with simple examples that this change makes worse; but I was not able to construct an example that this change makes better. There is probably one, but it is likely pretty convoluted.

**Comment by VSadov** [comment]
I have run the ArrayPool benchmarks ( https://github.com/dotnet/performance) with this change
Results: https://gist.github.com/VSadov/2c479e8bb553dabe26a79a24b8c5fa6a

The results are mostly inconclusive. There are small changes in either direction in both implementations of the pool (only one actually has the change). It seems to be just noise. It is not very surprising since most array pool operations do not involve allocations.

Basically this is a data point, but not very informative one.

**Comment by VSadov** [comment]
@jkotas - I am looking at tuning the perf of POH allocations per workitem https://github.com/dotnet/runtime/issues/13739

There are things that could be improved performance-wise, but to a limit. 
For example, SOH will always be better with shortlived/small objects. That is kind of by design.

I am wondering what scenarios you have found that this change makes worse? Perhaps that could lead to improvements.

**Comment by jkotas** [comment]
Here is the latest version of my ad-hoc test (that run slower with your change). It shows the kind of things I was experimenting with:
```
using System;
using System.Diagnostics;
using System.Buffers;

class Program
{
    static object[] gen2;

    static void CreateGen2()
    {
        gen2 = new object[10000000];
        for (int i = 0; i < gen2.Length; i++) gen2[i] = new object();
        GC.Collect(); 
    }

    static void Main(string[] args)
    {
        CreateGen2();

        for (;;)
        {
            Stopwatch sw = new Stopwatch();
            sw.Start();   
            for (int c = 0; c < 10000; c++)
            {
                byte[][] a = new byte[1000][];
                for (int i = 0; i < a.Length; i++) a[i] = ArrayPool<byte>.Shared.Rent(10);
                for (int j = 0; j < a.Length; j++) ArrayPool<byte>.Shared.Return(a[j]);
            }
            Console.WriteLine(sw.ElapsedMilliseconds.ToString());
        }
    }
}
```

**Comment by VSadov** [comment]
It looks like the PR generated good questions and discussion. 
However the benefits of the change itself are not convincing enough, so I am going to close this.
